### ISSUE TITLE
Loss moving to Nan/inf after epoch completion during fine-tuning titanet

### ISSUE THREAD

**Issue by abdulwahabamin** [NVIDIA/NeMo#5147]
Hi, I'm facing issues with finetuning titanet-large model. The loss goes to nan/inf in epoch 0 at the end and doesn't improve from there in the later epochs.

Epoch 0: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 150452/150452 [6:45:44<00:00,  6.18it/s, loss=nan, v_num=on_6]Epoch 0, global step 134943: 'val_loss' reached inf (best inf), saving model to '/voice/NeMo/examples/speaker_tasks/recognition/speaker_exps/default/version_6/checkpoints/default--val_loss=nan-epoch=0.ckpt' as top 3

My configuration file is:
`name: &name "TitaNet"
sample_rate: &sample_rate 8000

init_from_pretrained_model:
  speaker_tasks:
    name: 'titanet_large'
    include: ["preprocessor","encoder"]
    exclude: ["decoder.final"] # Add specific layer names here to exlude or just ["decoder"] if to exclude all of decoder pretrained weights

model:
  train_ds:
    manifest_filepath: ???
    sample_rate: 8000
    labels: null
    batch_size: 4
    max_duration: 16.7
    shuffle: True
    is_tarred: False
    tarred_audio_filepaths: null
    tarred_shard_strategy: "scatter"
    augmentor:
      noise:
        manifest_path: null
        prob: 0.6
        min_snr_db: -10
        max_snr_db: -10

      speed:
        prob: 0.3
        sr: *sample_rate
        resample_type: 'kaiser_fast'
        min_speed_rate: 0.95
        max_speed_rate: 1.05

      shift:
        prob: 0.6
        min_shift_ms: -5.0
        max_shift_ms: 5.0

      white_noise:
        prob: 0.5
        min_level: -90
        max_level: -46

  validation_ds:
    manifest_filepath: ???
    sample_rate: 8000
    labels: null
    batch_size: 4
    shuffle: False

  model_defaults:
    filters: 1024
    repeat: 3
    dropout: 0.1
    separable: true
    se: true
    se_context_size: -1
    kernel_size_factor: 1.0
    
  preprocessor:
    _target_: nemo.collections.asr.modules.AudioToMelSpectrogramPreprocessor
    normalize: "per_feature"
    window_size: 0.025
    sample_rate: *sample_rate
    window_stride: 0.01
    window: "hann"
    features: &n_mels 80
    n_fft: 512
    frame_splicing: 1
    dither: 0.00001

  spec_augment:
    _target_: nemo.collections.asr.modules.SpectrogramAugmentation
    freq_masks: 3
    freq_width: 4
    time_masks: 5
    time_width: 0.03

  encoder:
    _target_: nemo.collections.asr.modules.ConvASREncoder
    feat_in: *n_mels
    activation: relu
    conv_mask: true

    jasper:
      -   filters: ${model.model_defaults.filters}
          repeat: 1
          kernel: [3]
          stride: [1]
          dilation: [1]
          dropout: 0.0
          residual: false
          separable: ${model.model_defaults.separable}
          se: ${model.model_defaults.se}
          se_context_size: ${model.model_defaults.se_context_size}

      -   filters: ${model.model_defaults.filters}
          repeat:  ${model.model_defaults.repeat}
          kernel: [7]
          stride: [1]
          dilation: [1]
          dropout: ${model.model_defaults.dropout}
          residual: true
          separable: ${model.model_defaults.separable}
          se: ${model.model_defaults.se}
          se_context_size: ${model.model_defaults.se_context_size}

      -   filters: ${model.model_defaults.filters}
          repeat: ${model.model_defaults.repeat}
          kernel: [11]
          stride: [1]
          dilation: [1]
          dropout: ${model.model_defaults.dropout}
          residual: true
          separable: ${model.model_defaults.separable}
          se: ${model.model_defaults.se}
          se_context_size: ${model.model_defaults.se_context_size}

      -   filters: ${model.model_defaults.filters}
          repeat: ${model.model_defaults.repeat}
          kernel: [15]
          stride: [1]
          dilation: [1]
          dropout: ${model.model_defaults.dropout}
          residual: true
          separable: ${model.model_defaults.separable}
          se: ${model.model_defaults.se}
          se_context_size: ${model.model_defaults.se_context_size}

      -   filters: &enc_feat_out 3072
          repeat: 1
          kernel: [1]
          stride: [1]
          dilation: [1]
          dropout: 0.0
          residual: false
          separable: ${model.model_defaults.separable}
          se: ${model.model_defaults.se}
          se_context_size: ${model.model_defaults.se_context_size}

  decoder:
    _target_: nemo.collections.asr.modules.SpeakerDecoder
    feat_in: *enc_feat_out
    num_classes: 7205
    pool_mode: 'attention'
    emb_sizes: 192
    angular: True

  loss:
    scale: 30
    margin: 0.2

  optim:
    name: sgd
    lr: .006 #(original titanet-large was trained with 0.08 lr)
    weight_decay: 0.001

    # scheduler setup
    sched:
      name: CosineAnnealing
      warmup_ratio: 0.1
      min_lr: 0.0

trainer:
  devices: 2 # number of gpus (original titanet-large was trained on 4 nodes with 8 gpus each)
  max_epochs: 10
  max_steps: -1 # computed at runtime if not set
  num_nodes: 1
  precision: 16
  accelerator: gpu
  strategy: ddp
  deterministic: True
  enable_checkpointing: False
  logger: False
  log_every_n_steps: 1  # Interval of logging.
  val_check_interval: 1.0  # Set to 0.25 to check 4 times per epoch, or an int for number of iterations

exp_manager:
  exp_dir: null
  name: *name
  create_tensorboard_logger: True
  create_checkpoint_callback: True
`

**Comment by nithinraok** [comment]
Can you check if this fixes the nan issue? https://github.com/NVIDIA/NeMo/blob/a0dbc4ea3ee934088f409709415f5607c70247a5/nemo/collections/asr/modules/conv_asr.py#L876 

Also, I see you are using sample_rate = 8000, are you sure? And what is the lr? did you try decreasing lr

**Comment by abdulwahabamin** [comment]
I have updated the mentioned script. The training loss is not going toward nan now. I'm training it wth the original 0.08 learning rate right now. The loss remains around 14.8 after 2-3 epochs. Could I get some insights on the loss value from your end? I will try decreasing the learning rate in the next training iteration. 

and yes, the sampling rate for the audios I have is 8000 Hz.

**Comment by abdulwahabamin** [comment]
Hi, I went through the training cycle. The loss decreased to around 9.68 but didn't decrease any further. Upon testing it, I found out that the model is predicting one class only i.e 1 for similar speakers.

I might be going wrong somewhere, but I had a few basic questions that might help me understand how the speaker verification models are currently trained. As far as I have read on the models, they explain that they have used a combination of telephony and non-telephony data in the training set. Does this mean that the telephony data i.e 8khz speech was upsampled to 16khz during training? If yes, doesn't upsampling lead to some uncertain outputs? 

What do we do in cases when we have inference data that is only available in 8khz? Should we upsample it to 16khz or fine-tune it on the original 8khz files as I did? Keeping it to 8khz leads to undesirable effects as stated above and the model doesn't learn anything.

**Comment by nithinraok** [comment]
That is weird to observe. 

Did you change the sampling frequency in config yaml file?  https://github.com/NVIDIA/NeMo/blob/2574ddf71b3d488e544bba816da82678405e74c3/examples/speaker_tasks/recognition/conf/titanet-large.yaml#L2 

* Does this mean that the telephony data i.e 8khz speech was upsampled to 16khz during training? If yes, doesn't upsampling lead to some uncertain outputs? 

Yes, the narrow-band samples were upsampled and combined with wide band samples to train the models. This wont affect the outputs. During inference, you must upsample before passing to the model. 

* Should we upsample it to 16khz or fine-tune it on the original 8khz files as I did?  

I don't recommend fine-tuning the 16kHz model to 8Khz data since the initial features are learned over wide range of fft band. It will work but results might be unstable. To test this you might need to upsample and fine-tune, to see if you observe faster convergence.

**Comment by abdulwahabamin** [comment]
Yes, the sampling rate is 8000 as mentioned in the configuration shared above. As per our discussion so far, I'm assuming that I need to resample my data to 16khz during training and inference for maximum accuracy from titanet?

Also, I know this might be a basic question, but in terms of deploying models over a telephony network, I'll be resampling all my incoming audio to 16khz. Is 16khz only recommended because we get more features to predict on? If that is the case then in the resampling phase, we're just converting 8khz to 16khz i.e the features are just being expanded in terms of the original 8khz data. We get no additional features in this case. I'm slightly confused about the architecture decisions taken into consideration as the majority of SOTA models are trained on 16khz and not 8khz. Is this due to data scarcity for 8khz or some other factor?

**Comment by nithinraok** [comment]
Sorry missed this issue, 
There are 2 factors:
1. We don't want to train models for each sampling frequency, and 16kHz suits for most
2. As initial layers of the model learn for 16Khz (8 kHz Nyquist sampling rate) they learn high-level features at high mel bands. So adapting fine-tuned model wouldn't be as perfect as training from scratch for 8kHz data. Given you have enough data (approx 5 hrs per speaker) then better to train from scratch.

**Comment by ukemamaster** [comment]
Any update on this? 
@nithinraok I have 8khz telephony data. And i want to train the titanet-large model from scratch. Do i need to just change sampling rate in the config files? And nothing hard coded? What about window sizes for spectrograms? Do they adapt to any sampling rate? 

@abdulwahabamin Could you please share your experience?

**Comment by nithinraok** [comment]
You need to change sampling rate and related preprocessor values based on it. We haven;t trained for 8kHz. Our recommended way is to upsample 8kHz to 16kHz and use current config.
